### ISSUE TITLE
Assert vs. Exceptions vs. Caqti_error.t

### ISSUE THREAD

**Issue by jerben** [paurkedal/ocaml-caqti#43]
I had an issue when I ran into `assert_single_use` by using `Lwt.all ...` to run a list of queries on *a connection* where it just fails without error message. Would you accept a PR where I change that `assert` to throw an exception containing an error message that gives a hint, that it is not allowed to run multiple queries at the some time on one connection?

Would it even make sense maybe to get rid of all `assert`s and return `Caqti_error.t`?

**Comment by paurkedal** [comment]
Yes, a PR for raising a an exception with a more informative message would be appreciated. I suggest using `Failure` for this.

With respect to `Caqti_error.t`, this should not be used here.  Caqti uses the `result` type to indicate runtime errors, i.e. errors which are outside the control of the application, and which should be handled. On the other hand, failing the single-use test is a bug in the application, and should not be handled, other than to report it a such.
